packages/compat/src/transaction.ts

```diff
@@ -1,204 +1,30 @@
-import { AccountRole } from '@solana/instructions';
 import type { SignatureBytes } from '@solana/keys';
-import type { Blockhash } from '@solana/rpc-types';
-import {
-    appendTransactionInstruction,
-    createTransaction,
-    type IDurableNonceTransaction,
-    isAdvanceNonceAccountInstruction,
-    type ITransactionWithBlockhashLifetime,
-    type ITransactionWithFeePayer,
-    type ITransactionWithSignatures,
-    type Nonce,
-    setTransactionFeePayer,
-    setTransactionLifetimeUsingBlockhash,
-    setTransactionLifetimeUsingDurableNonce,
-    type Transaction,
-} from '@solana/transactions';
-import type {
-    MessageAccountKeys,
-    MessageCompiledInstruction,
-    PublicKey,
-    VersionedMessage,
-    VersionedTransaction,
-} from '@solana/web3.js';
-
-function convertAccount(
-    message: VersionedMessage,
-    accountKeys: MessageAccountKeys,
-    accountIndex: number,
-): IAccountMeta {
-    const accountPublicKey = accountKeys.get(accountIndex);
-    if (!accountPublicKey) {
-        throw new SolanaError(SOLANA_ERROR__TRANSACTION__ADDRESS_MISSING, {
-            index: accountIndex,
-        });
-    }
-    const isSigner = message.isAccountSigner(accountIndex);
-    const isWritable = message.isAccountWritable(accountIndex);
-
-    const role = isSigner
-        ? isWritable
-            ? AccountRole.WRITABLE_SIGNER
-            : AccountRole.READONLY_SIGNER
-        : isWritable
-          ? AccountRole.WRITABLE
-          : AccountRole.READONLY;
-
-    return {
-        address: accountPublicKey.toBase58() as Address,
-        role,
-    };
-}
-
-function convertInstruction(
-    message: VersionedMessage,
-    accountKeys: MessageAccountKeys,
-    instruction: MessageCompiledInstruction,
-): IInstruction {
-    const programAddressPublicKey = accountKeys.get(instruction.programIdIndex);
-    if (!programAddressPublicKey) {
-        throw new SolanaError(SOLANA_ERROR__TRANSACTION__ADDRESS_MISSING, {
-            index: instruction.programIdIndex,
-        });
-    }
-
-    const accounts = instruction.accountKeyIndexes.map(accountIndex =>
-        convertAccount(message, accountKeys, accountIndex),
+import { NewTransaction, type SignaturesMap, TransactionMessageBytes } from '@solana/transactions';
+import type { PublicKey, VersionedTransaction } from '@solana/web3.js';
+
+import { ReadonlyUint8Array } from '../../codecs-core/dist/types';
+
+function convertSignatures(transaction: VersionedTransaction, staticAccountKeys: PublicKey[]): SignaturesMap {
+    return Object.fromEntries(
+        transaction.signatures.map((sig, index) => {
+            const address = staticAccountKeys[index];
+            if (sig.every(b => b === 0)) {
+                // all-0 signatures are stored as null
+                return [address, null];
+            } else {
+                return [address, sig as ReadonlyUint8Array as SignatureBytes];
+            }
+        }),
```

Comment:
I think we want a length check here? To avoid `undefined` behavior. I know _we_ use the all-zeroes to create signature maps, but someone could have created this transaction another way, no?

Reply:
IIRC this all zeroes behaviour is the standard for setting null signatures on the old web3.js too.

Reply:
Yep, in old web3js we add an all-zero signature when a new signer is added, and modify those signatures in place when you sign. I'm happy to add a length check that `transaction.signatures` matches `transaction.message.header.numSigners` though, in case it's broken in some way.

Reply:
Added that, test here: https://github.com/solana-labs/solana-web3.js/blob/e12c4b765a498c37e3bd789e70379e16192c1e46/packages/compat/src/__tests__/transaction-test.ts#L201

Reply:
Works for me, thanks!